lib/google_contacts_api/api.rb

```diff
@@ -29,26 +29,101 @@ def get(link, params = {}, headers = {})
-    def post(link, params = {}, headers = {})
-      raise NotImplementedError
-      params["alt"] = "json"
-      @oauth.post("#{BASE_URL}#{link}?#{params.to_query}", headers)
+    def post_v2(link, params = {}, headers = {})
+      merged_params = params_with_defaults(params)
+      uri = "#{BASE_URL}#{link}?#{merged_params.to_query}"
+      begin
+        response = @oauth.post(uri, headers)
+      rescue => e
+        raise UnauthorizedError if defined?(e.response) && self.class.parse_response_code(e.response) == 401
+        raise e
+      end
     end
 
     # Put request to specified link, with query params
-    # Not tried yet
-    def put(link, params = {}, headers = {})
-      raise NotImplementedError
-      params["alt"] = "json"
-      @oauth.put("#{BASE_URL}#{link}?#{params.to_query}", headers)
+    def put_v2(link, params = {}, headers = {})
+      # Doesn't handle id yet
+      merged_params = params_with_defaults(params)
+      uri = "#{link}?#{merged_params.to_query}"
+      begin
+        response = @oauth.put(uri, headers)
+      rescue => e
+        raise UnauthorizedError if defined?(e.response) && self.class.parse_response_code(e.response) == 401
+        raise e
+      end
+    end
+
+    def post(contact)
+      options ={
+        headers: {
+          'Content-type' => 'application/json'
+          },
+        body: contact.entry_json
+      }
+      merged_params = params_with_defaults(options)
+      uri = URI.parse("https://www.google.com/m8/feeds/contacts/default/full/")
+
+      response = @oauth.post(uri, merged_params)
+    end
+
+    # Put request to specified link, with query params
+    def put(contact)
+      options ={
+        headers: {
+          'Content-type' => 'application/json',
+          'If-Match' => '*'
+          },
+        body: contact.entry_json
+      }
+      merged_params = params_with_defaults(options)
+
+      uri = URI.parse(contact.edit_link)
+      begin
+        response = @oauth.put(uri, merged_params)
+      rescue => e
+        GoogleContactsApi::Api.parse_response_code(e.response)
+      end
+    end
+
+    def post_group(group)
+      options ={
+        headers: {
+          'Content-type' => 'application/json'
+          },
+        body: group.entry_json
+      }
+      merged_params = params_with_defaults(options)
+      uri = URI.parse("https://www.google.com/m8/feeds/groups/default/full/")
+
+      response = @oauth.post(uri, merged_params)
+    end
+
+    def put_group(group)
+      options ={
+        headers: {
+          'Content-type' => 'application/json',
+          'If-Match' => '*'
+          },
+        body: group.entry_json
+      }
+      merged_params = params_with_defaults(options)
+
+      uri = URI.parse(group.edit_link)
+      begin
+        response = @oauth.put(uri, merged_params)
+        GoogleContactsApi::Api.parse_response_code(response)
+      rescue => e
+        GoogleContactsApi::Api.parse_response_code(e.response)
```

Comment:
This is a bit weird. If it works, you parse it. If it throws an error, you parse it. What if some other kind of error is thrown, where `e.response` will itself explode?